toplevel/identity/entity.go

```diff
@@ -0,0 +1,642 @@
+					"Required `id` attribute not found on alias element for entity id: %s", id))
+			}
+			aliasId := vals["id"].(string)
+			if _, exists := vals["name"]; !exists {
+				return nil, errors.New(fmt.Sprintf(
+					"Required `name` attribute not found on alias element for entity-alias id: %s", id))
+			}
+			aliasName := vals["name"].(string)
+			if _, exists := vals["mount_type"]; !exists {
+				return nil, errors.New(fmt.Sprintf(
+					"Required `mount_type` attribute not found on alias element for entity-alias id: %s", id))
+			}
+			mountType := vals["mount_type"].(string)
+			processedAliases = append(processedAliases, entityAlias{
+				Id:       aliasId,
+				Name:     aliasName,
+				AuthType: mountType,
+			})
+		}
+
+		processed = append(processed, entity{
+			Name:    name,
+			Id:      id,
+			Type:    "entity", // used for reconcile and output
+			Aliases: processedAliases,
+		})
+	}
+	return processed, nil
+}
+
+// performs concurrent requests to retrieve additional details for existing entities/entity aliases
+// these details require explicit requests to vault api for each entitiy/alias
+func getExistingEntitiesDetails(entities []entity, threadPoolSize int, customMetadataSupported bool) {
+	bwg := utils.NewBoundedWaitGroup(threadPoolSize)
+
+	for i := 0; i < len(entities); i++ {
+		bwg.Add(1)
+
+		go func(entity *entity) {
+			defer bwg.Done()
+			info := vault.GetEntityInfo(entity.Name)
+
+			if _, exists := info["metadata"]; !exists {
+				log.WithError(errors.New(fmt.Sprintf(
+					"Required `metadata` attribute not found for entity id: %s", entity.Id))).Fatal()
+			}
+			metadata := info["metadata"].(map[string]interface{})
+			if _, exists := info["policies"]; !exists {
+				log.WithError(errors.New(fmt.Sprintf(
+					"Required `policies` attribute not found for entity id: %s", entity.Id))).Fatal()
+			}
+			rawPolicies := info["policies"].([]interface{})
+			policies := []string{}
+			for _, policy := range rawPolicies {
+				policies = append(policies, policy.(string))
+			}
+
+			// TODO: make this a nested goroutine
+			for j := 0; j < len(entity.Aliases); j++ {
+				rawAlias := vault.GetEntityAliasInfo(entity.Aliases[j].Id)
+				if _, exists := rawAlias["mount_accessor"]; !exists {
+					log.WithError(errors.New(fmt.Sprintf(
+						"Required `mount_accessor` attribute not found for entity-alias id: %s", entity.Aliases[j].Id))).Fatal()
+				}
+				entity.Aliases[j].AccessorId = rawAlias["mount_accessor"].(string)
+				if customMetadataSupported {
+					// deviate from norm and do not fail on missing custom_metadata
+					if _, exists := rawAlias["custom_metadata"]; exists {
+						entity.Aliases[j].CustomMetadata = make(map[string]interface{})
+						aliasMap, ok := entity.Aliases[j].CustomMetadata.(map[string]interface{})
+						for k, v := range rawAlias["custom_metadata"].(map[string]interface{}) {
+							if ok {
+								aliasMap[k] = v
+							}
+						}
+					}
+				}
+			}
+
+			entity.Policies = policies
```

Comment:
Let's review the usage of policies on a review call on Monday. I'm sure this is OK, want to hear the flow from you.